pkg/imagestream/imagestream.go

```diff
@@ -0,0 +1,190 @@
+func Pull(apiClient *clients.Settings, name, nsname string) (*Builder, error) {
+	if apiClient == nil {
+		glog.V(100).Infof("The apiClient is empty")
+
+		return nil, fmt.Errorf("imageStream 'apiClient' cannot be empty")
+	}
+
+	err := apiClient.AttachScheme(imagev1.AddToScheme)
+	if err != nil {
+		glog.V(100).Info("Failed to add image v1 scheme to client schemes")
+
+		return nil, err
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the imageStream is empty")
+
+		return nil, fmt.Errorf("imageStream 'name' cannot be empty")
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the imageStream is empty")
+
+		return nil, fmt.Errorf("imageStream 'nsname' cannot be empty")
+	}
+
+	glog.V(100).Infof(
+		"Pulling imageStream object name %s from namespace %s", name, nsname)
+
+	builder := Builder{
+		apiClient: apiClient.Client,
+		Definition: &imagev1.ImageStream{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: nsname,
+			},
+			Spec: imagev1.ImageStreamSpec{},
+		},
+	}
+
+	if !builder.Exists() {
+		return nil, fmt.Errorf("imageStream object %s does not exist in namespace %s",
+			name, nsname)
+	}
+
+	builder.Definition = builder.Object
+
+	return &builder, nil
+}
+
+// Get fetches existing imageStream from cluster.
+func (builder *Builder) Get() (*imagev1.ImageStream, error) {
+	if valid, err := builder.validate(); !valid {
+		return nil, err
+	}
+
+	glog.V(100).Infof("Getting existing imageStream with name %s in namespace %s from cluster",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	imageStreamObj := &imagev1.ImageStream{}
+	err := builder.apiClient.Get(context.TODO(), goclient.ObjectKey{
+		Name:      builder.Definition.Name,
+		Namespace: builder.Definition.Namespace,
+	}, imageStreamObj)
+
+	if err != nil {
+		glog.V(100).Infof("imageStream object %s does not exist in namespace %s",
+			builder.Definition.Name, builder.Definition.Namespace)
+
+		return nil, err
+	}
+
+	return imageStreamObj, nil
+}
+
+// Exists checks whether the given imageStream exists.
+func (builder *Builder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof("Checking if imageStream %s exists in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// GetDockerImage fetches imageStream DockerImage value.
+func (builder *Builder) GetDockerImage() (string, error) {
+	if valid, err := builder.validate(); !valid {
+		return "", err
+	}
+
+	glog.V(100).Infof("Getting imageStream DockerImage value")
+
+	if !builder.Exists() {
+		return "", fmt.Errorf("imageStream object does not exist")
+	}
+
+	if len(builder.Object.Spec.Tags) == 0 {
+		return "", fmt.Errorf("imageStream object %s in namespace %s has no tags",
+			builder.Definition.Name, builder.Definition.Namespace)
+	}
+
+	return builder.Object.Spec.Tags[0].From.Name, nil
```

Comment:
It looks like From is a pointer and optional so it should have a check that it's not nil so that this function doesn't panic

Reply:
@elenagerman Could you still add the validation and leave in the len(Tags) == 0 validation? Even if it's not optional usually, the kubernetes api won't reject it so it's still possible to panic in edge cases which is undesirable

Reply:
yes, sure, done